<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" 
       name="vito">
       
  <!-- MODELS -->
  <xacro:include filename="$(find vito_description)/model/torso.urdf.xacro"/>
  <xacro:include filename="$(find vito_description)/model/table.urdf.xacro"/>
  <xacro:include filename="$(find vito_description)/model/materials.urdf"/>
  <xacro:include filename="$(find lwr_description)/model/kuka_lwr.urdf.xacro"/>
  <!-- <xacro:include filename="$(find soft_hand_description)/model/soft_hand.urdf.xacro"/> -->
  <!-- <xacro:include filename="$(find kit_head_description)/model/kit_head.urdf.xacro"/> -->
  <xacro:property name="width_ati" value="0.05" /> <!-- Square dimensions (widthxwidth) of beams -->
  <xacro:property name="mass_ati" value="0.0001" /> <!-- arbitrary value for mass -->


  <link name="world" />  
  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="vito_anchor"/>
  </joint>
  <link name="vito_anchor" />


  
  <!-- BODY -->
  <xacro:model_table name="table" 
                    parent="vito_anchor"
                    length="0.725"
                    width="1.05"
                    height="0.910"
                    plate_thickness="0.05">
    <origin xyz="-0.495 0.55 0" rpy="0 0 0"/>
  </xacro:model_table>

  <!-- WORKBENCH -->
  <xacro:model_table name="workbench"
                    parent="vito_anchor"
                    length="0.725"
                    width="1.25"
                    height="0.960"
                    plate_thickness="0.1">
  <origin xyz="-1.25 0.625 0.07" rpy="0 0 0"/>
  </xacro:model_table>
  <gazebo reference="workbench">
    <selfCollide>true</selfCollide>
</gazebo>

<!-- SFERA -->
<!--   <xacro:model_sphere name="sphere"
                      parent="vito_anchor"
                      radius="0.2">
    <origin xyz="-0.75 0 10" rpy="0 0 0"/>
  </xacro:model_sphere>
<gazebo reference="sphere">
    <selfCollide>true</selfCollide>
</gazebo> -->

  <!-- TORSO -->
  <xacro:model_torso name="torso" parent="vito_anchor">
    <origin xyz="0 0 0"/>
  </xacro:model_torso>
  
  <!-- LEFT ARM -->
  <xacro:kuka_lwr name="left_arm" parent="vito_anchor">
    <origin xyz="-0.165 -0.1085854 0.390" rpy="1.5707963267948966 0.5235987010713441 -0.5235990737060282"/>
  </xacro:kuka_lwr>

  <joint name="ati_left_joint" type="revolute">
    <parent link="left_arm_7_link"/>
    <child link="ati_left_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="0" upper="0" effort="0" velocity="0"/>
    <!-- <dynamics damping="0.7"/> -->

  </joint>
  <link name="ati_left_link">
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <box size="${width_ati} ${width_ati} ${width_ati}"/>
        </geometry>
      </collision>

      <visual>
        <origin xyz="0 0 ${width_ati/2}" rpy="0 0 0"/>
        <geometry>
            <box size="${width_ati} ${width_ati} ${width_ati}"/>
        </geometry>
        <material name="black"/>
      </visual>

      <inertial>
        <origin xyz="0 0 ${width_ati/2}" rpy="0 0 0"/>
        <mass value="${mass_ati}"/>
        <inertia
      ixx="${mass_ati / 12.0 * (width_ati*width_ati + width_ati*width_ati)}" ixy="0.0" ixz="0.0"
      iyy="${mass_ati / 12.0 * (width_ati*width_ati + width_ati*width_ati)}" iyz="0.0"
      izz="${mass_ati / 12.0 * (width_ati*width_ati + width_ati*width_ati)}"/>
      </inertial>
  </link>

  <!-- RIGHT ARM -->
  <xacro:kuka_lwr name="right_arm" parent="vito_anchor">
    <origin xyz="-0.165 0.10858 0.390" rpy="-1.5707963267948966 0.5235987010713441 0.5235990737060282"/>
  </xacro:kuka_lwr>


  <joint name="ati_right_joint" type="revolute">
    <parent link="right_arm_7_link"/>
    <child link="ati_right_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <limit lower="0" upper="0" effort="0" velocity="0"/>
    <!-- <dynamics damping="0.7"/> -->
  </joint>
  <link name="ati_right_link">
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <box size="${width_ati} ${width_ati} ${width_ati}"/>
        </geometry>
      </collision>

      <visual>
        <origin xyz="0 0 ${width_ati/2}" rpy="0 0 0"/>
        <geometry>
            <box size="${width_ati} ${width_ati} ${width_ati}"/>
        </geometry>
        <material name="black"/>
      </visual>

      <inertial>
        <origin xyz="0 0 ${width_ati/2}" rpy="0 0 0"/>
        <mass value="${mass_ati}"/>
        <inertia
      ixx="${mass_ati / 12.0 * (width_ati*width_ati + width_ati*width_ati)}" ixy="0.0" ixz="0.0"
      iyy="${mass_ati / 12.0 * (width_ati*width_ati + width_ati*width_ati)}" iyz="0.0"
      izz="${mass_ati / 12.0 * (width_ati*width_ati + width_ati*width_ati)}"/>
      </inertial>
  </link>

  <gazebo reference="ati_left_joint">
    <provideFeedback>true</provideFeedback>
  </gazebo>

  <gazebo reference="ati_right_joint">
    <provideFeedback>true</provideFeedback>
  </gazebo>

  <gazebo>
   <plugin name="force_sensor" filename="libgazebo_ros_ft_sensor.so">
    <updateRate>100.0</updateRate>
    <topicName>ati_left</topicName>
    <jointName>ati_left_joint</jointName>
  </plugin>
  </gazebo>

  <gazebo>
   <plugin name="force_sensor" filename="libgazebo_ros_ft_sensor.so">
    <updateRate>100.0</updateRate>
    <topicName>ati_right</topicName>
    <jointName>ati_right_joint</jointName>
  </plugin>
  </gazebo>


</robot>
